<launch>
    <!-- Task and logging arguments -->
    <arg name="task_type"/>
    <arg name="planner_logging_enabled"                 default="true"/>
    <arg name="controller_logging_enabled"              default="true"/>
    <arg name="test_id"                                 default="default"/>
    <arg name="task_max_time_override"                  default="false"/>
    <arg name="task_max_time"                           if="$(arg task_max_time_override)"/>
    <arg name="desired_motion_scale_factor_override"    default="false"/>
    <arg name="desired_motion_scale_factor"             if="$(arg desired_motion_scale_factor_override)"/>

    <!-- Simulator arguments -->
    <arg name="launch_simulator"                        default="true"/>
    <arg name="feedback_covariance"                     default="0.0"/>
    <arg name="start_bullet_viewer"                     default="true"/>
    <arg name="screenshots_enabled"                     default="false"/>

    <!-- Planner trial type -->
    <arg name="planner_trial_type"                      default="multi_model_accuracy_test"/>
    <!--<arg name="planner_trial_type"                      default="multi_model_controller_test"/>-->
    <!--<arg name="planner_trial_type"                      default="constraint_single_model_constraint_controller"/>-->

    <!-- Planner visualization arguments -->
    <arg name="disable_all_visualizations"              default="false"/>
    <arg name="visualize_object_desired_motion"         default="true"/>
    <arg name="visualize_gripper_motion"                default="true"/>
    <arg name="visualize_object_predicted_motion"       default="true"/>
    <arg name="visualize_prm"                           default="false"/>
    <arg name="visualize_free_space_graph"              default="false"/>
    <arg name="visualize_correspondences"               default="false"/>

    <!-- Model/Controller arguments -->
    <!-- Diminishing rigidity -->
    <arg name="deformability_override"                  default="false"/>
    <arg name="translational_deformability"             if="$(arg deformability_override)"/>
    <arg name="rotational_deformability"                if="$(arg deformability_override)"/>
    <!-- Adaptive Jacobian -->
    <arg name="adaptive_model_learning_rate"            default="0.000001"/>
    <!-- Directional rigidity -->
    <arg name="translational_dir_deformability"         default="4.0"/>
    <arg name="translational_dis_deformability"         default="10.0"/>
    <arg name="rotational_dis_deformability"            default="10.0"/>
    <!-- Stretching Avoidance Controller -->
    <arg name="stretching_cosine_threshold_override"    default="false"/>
    <arg name="stretching_cosine_threshold"             if="$(arg stretching_cosine_threshold_override)"/>
    <!-- Multi-model Bandit  -->
    <arg name="kalman_parameters_override"              default="false"/>
    <arg name="process_noise_factor"                    if="$(arg kalman_parameters_override)"/>
    <arg name="observation_noise_factor"                if="$(arg kalman_parameters_override)"/>
    <arg name="bandit_algorithm"                        default="UCB"/>
    <arg name="correlation_strength_factor_override"    default="false"/>
    <arg name="correlation_strength_factor"             if="$(arg correlation_strength_factor_override)"/>

    <!-- Planner internal arguments -->
    <arg name="launch_planner"                          default="true"/>
    <arg name="collect_results_for_all_models"          default="true"/>
    <arg name="use_random_seed"                         default="false"/>
    <arg name="static_seed_override"                    default="false"/>
    <arg name="static_seed"                             if="$(arg static_seed_override)"/>

    <!-- Setup task parameters -->
    <include file="$(find deformable_manipulation_experiment_params)/launch/$(arg task_type)_params.launch">
        <arg name="stretching_cosine_threshold"         value="$(arg stretching_cosine_threshold)"  if="$(arg stretching_cosine_threshold_override)"/>
        <arg name="desired_motion_scale_factor"         value="$(arg desired_motion_scale_factor)"  if="$(arg desired_motion_scale_factor_override)"/>
    </include>
    <param name="log_folder"                            value="$(find smmap)/logs/$(arg task_type)/$(arg test_id)/"         type="string"/>

    <!-- Setup the simulator -->
    <group ns="deform_simulator_node">
        <param name="feedback_covariance"               value="$(arg feedback_covariance)"      type="double"/>
        <param name="start_bullet_viewer"               value="$(arg start_bullet_viewer)"      type="bool"/>
        <param name="screenshots_enabled"               value="$(arg screenshots_enabled)"      type="bool"/>
        <param name="settle_time"                       value="4.0"/>
    </group>
    <env name="OSG_FILE_PATH" value="$(find OpenSceneGraph)/data"/>
    <node name="deform_simulator_node" pkg="deform_simulator" type="custom_scene_node" required="true" if="$(arg launch_simulator)"/>


    <!-- Setup the planner -->
    <group ns="smmap_planner_node">
        <param name="planner_trial_type"                value="$(arg planner_trial_type)"                   type="string"/>
        <param name="use_random_seed"                   value="$(arg use_random_seed)"                      type="bool"/>
        <param name="static_seed"                       value="$(arg static_seed)"                          type="string"   if="$(arg static_seed_override)"/>

        <param name="planner_logging_enabled"           value="$(arg planner_logging_enabled)"              type="bool"/>
        <param name="controller_logging_enabled"        value="$(arg controller_logging_enabled)"           type="bool"/>

        <!--<param name="use_stored_world_state"                value="true"                                    type="bool"/>-->
        <param name="use_stored_world_state"                value="false"                                   type="bool"/>
        <param name="world_state_file_name_prefix"          value="world_state"                             type="string"/>
        <!--<param name="world_state_file_name_suffix_to_load"  value="cant_find_feasible"                type="string"/>-->
        <param name="world_state_file_name_suffix_to_load"  value="new_start_position_rotating_wrong_way"                type="string"/>

        <param name="disable_all_visualizations"        value="$(arg disable_all_visualizations)"           type="bool"/>
        <param name="visualize_object_desired_motion"   value="$(arg visualize_object_desired_motion)"      type="bool"/>
        <param name="visualize_gripper_motion"          value="$(arg visualize_gripper_motion)"             type="bool"/>
        <param name="visualize_object_predicted_motion" value="$(arg visualize_object_predicted_motion)"    type="bool"/>
        <param name="visualize_prm"                     value="$(arg visualize_prm)"                        type="bool"/>
        <param name="visualize_free_space_graph"        value="$(arg visualize_free_space_graph)"           type="bool"/>
        <param name="visualize_correspondences"         value="$(arg visualize_correspondences)"            type="bool"/>

        <group ns="rrt">
            <param name="timeout"                       value="1200.0"                      type="double"/>
            <param name="reuse_old_results"             value="false"                       type="bool"/>
            <param name="store_new_results"             value="true"                        type="bool"/>
        </group>

        <group ns="prm">
            <param name="num_nearest"                   value="5"                           type="int"/>
            <param name="num_samples"                   value="1000"                        type="int"/>
        </group>

        <group ns="band">
            <param name="logging_enabled"               value="$(arg planner_logging_enabled)"                              type="bool"/>
            <param name="log_folder"                    value="$(find smmap)/logs/$(arg task_type)/$(arg test_id)/"         type="string"/>
            <param name="use_stored_band"               value="false"                                                       type="bool"/>
            <param name="band_file_name_prefix"         value="band"                                                        type="string"/>
            <param name="band_file_name_suffix_to_load" value="2017-10-30__00-12-51-043"                                    type="string"/>
        </group>

        <!-- Dinimishing Rigidity parameters and overrides -->
        <group ns="diminishing_rigidity">
            <param name="translational_deformability"   value="$(arg translational_deformability)"      type="double"   if="$(arg deformability_override)"/>
            <param name="rotational_deformability"      value="$(arg rotational_deformability)"         type="double"   if="$(arg deformability_override)"/>
        </group>

        <!-- Adaptive Jacobian parameters and overrides -->
        <group ns="adaptive_model">
            <param name="adaptive_model_learning_rate"  value="$(arg adaptive_model_learning_rate)"     type="double"/>
        </group>

        <!-- Constraint Model parameters and overrides -->
        <group ns="constraint_model">
            <param name="translational_dir_deformability"   value="$(arg translational_dir_deformability)"  type="double"/>
            <param name="translational_dis_deformability"   value="$(arg translational_dis_deformability)"  type="double"/>
            <param name="rotational_dis_deformability"      value="$(arg rotational_dis_deformability)"     type="double"/>
        </group>

        <!-- Multi-model parameters and overrides -->
        <group ns="multi_model">
            <param name="process_noise_factor"              value="$(arg process_noise_factor)"             type="double"   if="$(arg kalman_parameters_override)"/>
            <param name="observation_noise_factor"          value="$(arg observation_noise_factor)"         type="double"   if="$(arg kalman_parameters_override)"/>
            <param name="collect_results_for_all_models"    value="$(arg collect_results_for_all_models)"   type="bool"/>
            <param name="correlation_strength_factor"       value="$(arg correlation_strength_factor)"      type="double"   if="$(arg correlation_strength_factor_override)"/>
            <param name="reward_scale_annealing_factor"     value="0.9"                                     type="double"/>
        </group>

        <!-- Old pure Jacobian based controller -->
        <group ns="jacobian_controller">
            <param name="optimization_enabled"              value="true"                                    type="bool"/>
        </group>

        <!-- stretching_avoidance_controller arguments -->
        <group ns="stretching_avoidance_controller">
            <param name="solver_type"                   value="gradient_descent"                        type="string"/>
            <param name="max_sampling_counts"           value="5000"                                    type="int"/>
        </group>
    </group>
    <node name="smmap_planner_node" pkg="smmap" type="planner_test_node_$(arg bandit_algorithm)" required="true" if="$(arg launch_planner)"/>
</launch>
